controlplane/kubeadm/controllers/controller.go

```diff
@@ -442,21 +449,43 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
 	return nil
 }
 
-// reconcileHealth performs health checks for control plane components and etcd
+// reconcileControlPlaneHealth performs health checks for control plane components and etcd
 // It removes any etcd members that do not have a corresponding node.
 // Also, as a final step, checks if there is any machines that is being deleted.
-func (r *KubeadmControlPlaneReconciler) reconcileHealth(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, controlPlane *internal.ControlPlane) (ctrl.Result, error) {
+func (r *KubeadmControlPlaneReconciler) reconcileControlPlaneHealth(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, controlPlane *internal.ControlPlane) (ctrl.Result, error) {
+	logger := r.Log.WithValues("namespace", kcp.Namespace, "kubeadmControlPlane", kcp.Name)
+
+	// If there is no KCP-owned control-plane machines, then control-plane has not been initialized yet.
+	if controlPlane.Machines.Len() == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	for i := range controlPlane.Machines {
+		m := controlPlane.Machines[i]
+		// Initialize the patch helper.
+		patchHelper, err := patch.NewHelper(m, r.Client)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		defer func() {
+			// Always attempt to Patch the Machine conditions after each health reconciliation.
+			if err := patchHelper.Patch(ctx, m); err != nil {
```

Comment:
Machine remediation will call `reconcileControlPlaneHealth`, right? If it gets called then the conditions will be updated.

It doesn't cover delete workflow because we don't call `reconcileControlPlaneHealth`, and hence these conditions will not be updated. During deletion Machines will signal as deleting and `NodeHealthy` condition will signal the deletion so it may be unnecessary to do a health check on pods during deletion.

Reply:
> Machine remediation will call reconcileControlPlaneHealth, right

I'm not sure we will call reconcileControlPlaneHealth during remediation because remediation should happen before we are checking  reconcileControlPlaneHealth (also, curently we are returning if cluster is unhealthy)

> It doesn't cover delete workflow because... so it may be unnecessary to do a health check on pods during deletion

This is why I was suggesting to consider have this code in updateStatus

Reply:
Added reconcileHealth to reconcile delete with the minimum changes possible.
Ideally, `TargetClusterEtcdIsHealthy` and `TargetClusterControlPlaneIsHealthy` could be removed totally, but did not do that in this PR.

Reply:
Had to remove error checks for `TargetClusterEtcdIsHealthy` and  `TargetClusterControlPlaneIsHealthy` during reconcileDelete because control plane may be unreachable during delete.